pkg/appliance/functions.go

```diff
@@ -776,3 +788,251 @@ func PrettyBytes(v float64) string {
 	}
 	return fmt.Sprintf("%.2fYB", v)
 }
+
+type ManifestConfig struct {
+	MediaType string
+	Size      int
+	Digest    string
+}
+type DockerManifest struct {
+	SchemaVersion int
+	MediaType     string
+	Config        ManifestConfig
+	Layers        []ManifestConfig
+}
+
+type ImageJSON struct {
+	Image string `json:"image"`
+}
+
+func DownloadDockerBundles(ctx context.Context, out io.Writer, client *http.Client, zipName, registry string, images map[string]string, ciMode bool) (*os.File, fs.FileInfo, error) {
+	// Create zip-archive
+	archive, err := os.Create(fmt.Sprintf("/tmp/%s.zip", zipName))
+	if err != nil {
+		return nil, nil, err
+	}
+	defer archive.Close()
+	zipWriter := zip.NewWriter(archive)
+	defer zipWriter.Close()
+	fileStat, err := archive.Stat()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	log.Info("downloading image layers for ", zipName)
+	var p *tui.Progress
+	if !ciMode {
+		p = tui.New(ctx, out)
+		defer p.Wait()
+	}
+	errChan := make(chan error)
+	var wg sync.WaitGroup
+	wg.Add(len(images))
+	for image, tag := range images {
+		go downloadDockerImageBundle(ctx, out, client, ciMode, zipWriter, registry, image, tag, p, errChan, &wg)
+	}
+
+	go func(wg *sync.WaitGroup, errChan chan error) {
+		wg.Wait()
+		close(errChan)
+	}(&wg, errChan)
+
+	var errs *multierror.Error
+	for err := range errChan {
+		if err != nil {
+			errs = multierror.Append(errs, err)
+		}
+	}
+	log.Info("docker layers downloaded")
+
+	return archive, fileStat, errs.ErrorOrNil()
+}
+
+func downloadDockerImageBundle(ctx context.Context, out io.Writer, client *http.Client, ciMode bool, zipWriter *zip.Writer, registry, image, tag string, p *tui.Progress, mainErrChan chan<- error, wg *sync.WaitGroup) {
+	defer wg.Done()
+	headers := map[string]string{
+		"Accept":       "application/vnd.docker.distribution.manifest.v2+json",
+		"Content-Type": "application/json",
+	}
+
+	// Download mainfest.json
+	manifestURL := fmt.Sprintf("%s/v2/%s/manifests/%s", registry, image, tag)
+	manifestReq, err := http.NewRequestWithContext(ctx, http.MethodGet, manifestURL, nil)
+	if err != nil {
+		mainErrChan <- err
+		return
+	}
+	for k, v := range headers {
+		manifestReq.Header.Set(k, v)
+	}
+	manifestRes, err := client.Do(manifestReq)
+	if manifestRes.StatusCode != http.StatusOK || err != nil {
```

Comment:
manifestRes can be null, should be enough to only check err
```suggestion
	if err != nil {
```

Reply:
same as configRes here